src/npctrade.cpp

```diff
@@ -612,50 +663,196 @@ bool trading_window::perform_trade( npc &np, const std::string &deal )
+
+                    while( ip_container->parent ) {
+                        // Found the top-most selected container
+                        if( ip_container->parent->selected ) {
+                            ip_container = ip_container->parent;
+                            ip_container->selected = false;
+                            ip_container->marked = false;
+
+                            adjust_balance( *ip_container, np, 1 );
+                            break;
+                        } else if( !ip_container->parent->selected && ip_container->parent->marked ) {
+                            ip_container = ip_container->parent;
+                            ip_container->marked = false;
+                        } else {
+                            break;
+                        }
+                    }
+
+                    // Because the @item_pricing of the container includes price and weigth of the individual items,
+                    // after deselecting the container itself, go through once more and add the @item_pricing of the remaining selected items.
+                    // We want to find orphaned containers, that are marked for trade, and we want to SELECT them too. If we treat them as regular items,
+                    // their price will be used to calculate the balance, and then the price of their contents is going to be duplicately calculated, leading to problems.
+                    std::queue<item_pricing *> nodes;
+                    nodes.push( ip_container );
+
+                    while( nodes.size() ) {
+                        item_pricing *current = nodes.front();
+
+                        for( item_pricing *ip2 : current->contents ) {
+                            // We found an orphaned subcontainer. SELECT it for trading. Its contents should be already marked
+                            if( ip2->is_container && ip2->marked && ip2->contents.size() ) {
+                                ip2->selected = true;
+                                adjust_balance( *ip2, np, 1 );
+                            }
+                            // regular item or container without contents. Treat as a regular item, i.e select for trading and add price
+                            else if( ip2->marked ) {
+                                ip2->selected = true;
+                                ip2->marked = true;
+                                adjust_balance( *ip2, np, get_change_amount( *ip2, false ) );
+                            }
+                            // unmarked container. we have to make sure that we check it for orphaned unselected items or orphaned containers
+                            else if( !ip2->marked && ip2->contents.size() ) {
+                                nodes.push( ip2 );
+                            }
+
+                        }
+                        nodes.pop();
+                    }
+                    continue;
                 }
```

Comment:
this whole while loop needs to be a separate function. off hand i suspect it would be slightly better as a recursive function but that implementation detail doesn't bother me at this point